<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Mapping to a reference genome</title>

<script src="site_libs/header-attrs-2.25/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">tirupati_genomics</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/markravinet/tirupati_genomics">
    <span class="fab fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Mapping to a reference genome</h1>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-01-22
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 2
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>tirupati_genomics/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcommarkravinettirupatigenomicstree3f129fa1e4eb0a6e9372205c7b1edd4aa827b8a0targetblank3f129faa">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/markravinet/tirupati_genomics/tree/3f129fa1e4eb0a6e9372205c7b1edd4aa827b8a0" target="_blank">3f129fa</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcommarkravinettirupatigenomicstree3f129fa1e4eb0a6e9372205c7b1edd4aa827b8a0targetblank3f129faa"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/markravinet/tirupati_genomics/tree/3f129fa1e4eb0a6e9372205c7b1edd4aa827b8a0" target="_blank">3f129fa</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/

Untracked files:
    Untracked:  website_setupt.R

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown
(<code>analysis/3_Mapping_to_a_reference_genome.Rmd</code>) and HTML
(<code>docs/3_Mapping_to_a_reference_genome.html</code>) files. If
you’ve configured a remote Git repository (see
<code>?wflow_git_remote</code>), click on the hyperlinks in the table
below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/markravinet/tirupati_genomics/blob/3f129fa1e4eb0a6e9372205c7b1edd4aa827b8a0/analysis/3_Mapping_to_a_reference_genome.Rmd" target="_blank">3f129fa</a>
</td>
<td>
Mark Ravinet
</td>
<td>
2024-01-22
</td>
<td>
update files
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>One of the first steps we need to take along our pathway to
population/speciation genomic analyses is mapping our data to a
reference genome.</p>
<p>Using alignment software, we essentially find where in the genome our
reads originate from and then once these reads are aligned, we are able
to either call variants or construct a consensus sequence for our set of
aligned reads.</p>
<div id="getting-access-to-the-reference-genome" class="section level3">
<h3>Getting access to the reference genome</h3>
<p>We will be aligning our sequence data to the <em>Passer
domesticus</em> reference genome, first published by <a
href="https://www.science.org/doi/epdf/10.1126/sciadv.1602996">Elgvin
<em>et al.</em> (2017)</a>.</p>
<p>You will find a copy of the reference genome in the
<code>/home/data/reference/</code> directory. Make a
<code>reference</code> directory in your own home directory and then
copy the reference genome like so:</p>
<pre class="shell"><code>cd ~
mkdir reference
cd reference
cp /home/data/reference/house_sparrow_genome_assembly-18-11-14_masked.fa .</code></pre>
<p>In order to align reads to the genome, we are going to use
<code>bwa</code> which is a very fast and straightforward aligner. See
<a href="http://bio-bwa.sourceforge.net/">here</a> for more details on
<code>bwa</code>.</p>
<div id="indexing-the-reference-genome" class="section level4">
<h4>Indexing the reference genome</h4>
<p>Before we can actually perform an alignment, we need to index the
reference genome we just copied to our home directories. This
essentially produces an index for rapid searching and aligning. We use
the <code>bwa index</code> tool to achieve this. However, the command
takes some time to run, so we will use a useful Unix utility called
<code>screen</code> to run it in the background (we will return to
<strong>why</strong> it is advantageous to use <code>screen</code>
momentarily).</p>
<p>Use the following command to start a screen:</p>
<pre class="shell"><code>screen -S index</code></pre>
<p>The terminal should flicker and you will now be in another screen.
You can prove this to yourself with the following command:</p>
<pre class="shell"><code>screen -ls</code></pre>
<p>This will list the running screens and should show you one called
<code>index</code>. If you are inside this screen, you will also see
that it is listed as attached.</p>
<p>Now that you are sure you are inside the screen, run the following
command.</p>
<pre class="shell"><code>bwa index house_sparrow_genome_assembly-18-11-14_masked.fa</code></pre>
<p>Let’s breakdown what we actually did here. As you might imagine
<code>bwa index</code> is a tool for indexing. We simply specify the
reference fasta file from which to build our genome index.</p>
<p>As you will have noticed by now, this command takes quite some time
to run. Rather than sit there and wait, you can press
<code>ctrl + ad</code> keys all together to shift back to the main
screen (i.e. the one you started in).</p>
<div id="why-screen" class="section level5">
<h5>Why screen?</h5>
<p>While we are waiting for the index to complete, this is a good moment
to discuss the benefits of <code>screen</code>. You’ve already seen that
screen allows you to run processes in the background, but in addition to
that, screen also protects your session. If you are running an analysis
and the terminal you are in crashes or your computer stops or something,
it will send a kill signal to your analysis and stop it. However, if you
have placed that analysis in another screen, this kill signal will not
reach it - meaning you can safely log out and log back in to a node and
your work will be running. This is just one way to run an analysis that
will take a long time.</p>
<p>To return to your screen at any time and check the analysis, use the
following:</p>
<pre class="shell"><code>screen -rd</code></pre>
<p>If you have multiple screens, you will need to specify the ID number
- to discover this, use <code>screen -ls</code>.</p>
</div>
<div id="back-to-the-indexing" class="section level5">
<h5>Back to the indexing</h5>
<p>OK so hopefully by now, the indexing should have completed - on our
cluster it takes about 15 mins. If not, then kill the process using
<code>ctrl + c</code> and exit the screen by typing
<code>exit</code>.</p>
<p>To save time and to allow everyone to proceed at the same pace, we
will instead copy some already built indexes to our home directory</p>
<pre class="shell"><code>cp /home/data/reference/house_sparrow_genome_assembly-18-11-14_masked.fa* .</code></pre>
<p>Use <code>ls</code> to take a look, but this will have copied in
about 5 files all with the <code>P_nyererei_v2.fasta.</code> prefix that
we will use for a reference alignment.</p>
<p>When <code>bwa</code> aligns reads, it needs access to these files,
so they should be in the same directory as the reference genome. Then
when we actually run the alignment, we tell <code>bwa</code> where the
reference is and it does the rest. To make this easier, we will make a
variable pointing to the reference.</p>
<pre class="shell"><code>REF=~/reference/house_sparrow_genome_assembly-18-11-14_masked.fa</code></pre>
</div>
</div>
<div id="performing-a-paired-end-alignment" class="section level4">
<h4>Performing a paired end alignment</h4>
<p>Now we are ready to align our sequences! To simplify matters, we will
first try this on a single individual. First, we will create a directory
to hold our aligned data:</p>
<pre class="shell"><code>cd ~
mkdir align
cd align</code></pre>
<p>As a side note, it is good practice to keep your data well organised
like this, otherwise things can become confusing and difficult in the
future.</p>
<p>To align our individual we will use <code>bwa</code>. You might want
to first have a look at the options available for it simply by calling
<code>bwa</code>. We are actually going to use <code>bwa mem</code>
which is the best option for short reads.</p>
<p>As we mentioned above, we will use the individual -
<code>10558.PunPundMak</code> which we have already trimmed. There are
two files for this individual - <code>R1</code> and <code>R2</code>
which are forward and reverse reads respectively.</p>
<p>Let’s go ahead and align our data, we will break down what we did
shortly after. Note that we run this command from the home
directory.</p>
<pre class="shell"><code>bwa mem -M -t 4 $REF \
/home/data/reads/PDOM2023IND0001J_30707403185-HS1_S178_L002_R1_001.fastq.gz \
/home/data/reads/PDOM2023IND0001J_30707403185-HS1_S178_L002_R2_001.fastq.gz &gt; PDOM2023IND0001J.sam</code></pre>
<p>Since we are only using a shortened fastq file, with 100K reads in
it, this should just take a couple of minutes. In the meantime, we can
breakdown what we actually did here.</p>
<ul>
<li><code>-M</code> is a standard flag that tells <code>bwa</code> to
mark any low quality alignments (i.e. split across long distances) as
secondary - we need this for downstream compatability.</li>
<li><code>-t</code> tells <code>bwa</code> how many threads (cores) on a
cluster to use - this effectively determines its speed.</li>
<li>Following these options, we then specify the reference genome, the
forward reads and the reverse reads. Finally we write the output of the
alignment to a <strong>SAM file</strong>.</li>
</ul>
<p>Once your alignment has ended, you will see some alignment statistics
written to the screen. We will come back to these later - first, we will
learn about what a SAM file actually is.</p>
</div>
<div id="sam-files" class="section level4">
<h4>SAM files</h4>
<p>Lets take a closer a look at the output. To do this we will use
<code>samtools</code>. More details on <code>samtools</code> can be
found <a href="http://www.htslib.org/doc/samtools.html">here</a></p>
<pre class="shell"><code>cd align
samtools view -h PDOM2023IND0001J.sam | head
samtools view PDOM2023IND0001J.sam | head</code></pre>
<p>This is a SAM file - or sequence alignment/map format. It is
basically a text format for storing an alignment. The format is made up
of a header where each line begins with <code>@</code>, which we saw
when we used the <code>-h</code> flag and an alignment section.</p>
<p>The header gives us details on what has been done to the SAM, where
it has been aligned and so on. We will not focus on it too much here but
there is a very detailed SAM format specification <a
href="https://samtools.github.io/hts-specs/SAMv1.pdf">here</a>.</p>
<p>The alignment section is more informative at this stage and it has at
least 11 fields. The most important for now are the first four. Take a
closer look at them.</p>
<pre class="shell"><code>samtools view PDOM2023IND0001J.sam | head | cut -f 1-4</code></pre>
<p>Here we have:</p>
<ul>
<li>The sequence ID</li>
<li>The flag - these have various meanings, 0 = mapped, 4 =
unmapped</li>
<li>Reference name - reference scaffold the read maps to</li>
<li>Position - the mapping position/coordinate for the read</li>
</ul>
<p><strong>Note!!</strong> The other fields are important, but for our
purposes we will not examine them in detail here.</p>
<p>Now, lets look at the mapping statistics again:</p>
<pre class="shell"><code>samtools flagstat PDOM2023IND0001J.sam</code></pre>
<p>This shows us that a total of 200k reads were read in (forward and
reverse), that around 94% mapped successfully, 81% mapped with their
mate pair, 1.94% were singletons and the rest did not map.</p>
</div>
<div id="bam-files" class="section level4">
<h4>BAM files</h4>
<p>One problem with SAM files is that for large amounts of sequence
data, they can rapidly become HUGE in size. As such, you will often see
them stored as BAM files (Binary Aligment/Mapping format). A lot of
downstream analyses require the BAM format so our next task is to
convert our SAM to a BAM.</p>
<pre class="shell"><code>samtools view PDOM2023IND0001J.sam -b -o PDOM2023IND0001J.bam</code></pre>
<p>Here the <code>-b</code> flag tells <code>samtools</code> to output a
BAM and <code>-o</code> identifies the output path. Take a look at the
files with <code>ls -lah</code> - you will see a substantial difference
in their size. This would be even more striking if we were to map a full
dataset.</p>
<p>You can view bamfiles in the same way as before.</p>
<pre class="shell"><code>samtools view PDOM2023IND0001J.bam | head</code></pre>
<p>Note that the following will not work (although it does for SAM)
because it is a binary format</p>
<pre class="shell"><code>head PDOM2023IND0001J.bam</code></pre>
<p>Before we can proceed to more exciting analyses, there is one more
step we need to perform - sorting the bamfile. Most downstream analyses
require this in order to function efficiently</p>
<pre class="shell"><code>samtools sort PDOM2023IND0001J.bam -o PDOM2023IND0001J_sort.bam</code></pre>
<p>Once this is run, we will have a sorted bam. One point to note here,
could we have done this is a more efficient manner? The answer is yes,
actually we could have run all of these commands in a single line using
pipes like so:</p>
<pre class="shell"><code>bwa mem -M -t 4 $REF /home/data/reads/PDOM2023IND0001J_30707403185-HS1_S178_L002_R1_001.fastq.gz /home/data/reads/PDOM2023IND0001J_30707403185-HS1_S178_L002_R1_001.fastq.gz | samtools view -b | samtools sort -T PDOM2023IND0001J &gt; ./align/PDOM2023IND0001J_sort.bam</code></pre>
<p>However as you may have noticed, we have only performed this on a
single individual so far… what if we want to do it on multiple
individuals? Do we need to type all this everytime? The answer is no -
we could do this much more efficiently using the bash scripting tools we
learned earlier today.</p>
</div>
</div>
<div id="working-with-multiple-individuals" class="section level3">
<h3>Working with multiple individuals</h3>
<p>There are actually multiple ways to work with multiple individuals.
One way is to use a bash script to loop through all the files and align
them one by one. We will examine this way in detail together. An
alternative way is to use some form of parallelisation and actually map
them all in parallel. We will also demonstrate an example of this, but
it is quite advanced and is really only to give you some familiarity
with the approach.</p>
<div id="using-a-bash-script-to-loop-through-and-align-individuals"
class="section level4">
<h4>Using a bash script to loop through and align individuals</h4>
<p>We are going to work together to write a bash script which you can
use to map multiple individuals one by one. It is typically a lot more
straightforward to write a script locally on your machine and then
upload it to the cluster to make sure it does what you want. There are
plenty of different scripting and code specific text editors out there
but a free, multi-platform, open source editor we recommomend is <a
href="https://atom.io/">Atom</a>.</p>
<p>The first thing we will do is initiate the script with the line
telling the interpreter that the file must be treated as a bash
script:</p>
<pre class="shell"><code>#!/bin/sh</code></pre>
<p>Next, we declare some variables in our script. As we learned from our
adventures with <code>bwa mem</code> above, we need to point to the
reference genome. We do this like so:</p>
<pre class="shell"><code>REF=~/reference/house_sparrow_genome_assembly-18-11-14_masked.fa</code></pre>
<p>Next, we will declare an array to ensure that we have all our
individuals</p>
<pre class="shell"><code>INDS=($(for i in /home/data/reads/*_R1_001.fastq.gz; do echo $(basename ${i%_R*}); done))</code></pre>
<p>As we learned before, this will create a list of individuals which we
can then loop through in order to map each individual. Here we used bash
substitution to take each forward read name, remove the directory and
leave only the individual name.</p>
<p>If you want to, decleare the array in your command line and then test
it (i.e. type <code>echo ${IND[@]}</code>). You will see that we have
only individual names, which gives us some flexibility to take our
individual name and edit it inside our <code>for</code> loop (i.e. it
makes defining input and output files much easier.</p>
<p>Next we will add the actual <code>for</code> loop to our script. We
will use the following:</p>
<pre class="shell"><code>for IND in ${INDS[@]};
do
    # declare variables
    FORWARD=/home/data/reads/${IND}_R1_001.fastq.gz
    REVERSE=/home/data/reads/${IND}_R2_001.fastq.gz
    OUTPUT=~/align/${INDS%%_*}_sort.bam

done</code></pre>
<p>In this first version of our loop, we are making the
<code>$FORWARD</code>, <code>$REVERSE</code> and <code>$OUTPUT</code>
variables, making this much easier for us to edit later. It is a good
idea to test the variables we are making in each loop, using the
<code>echo</code> command. Feel free to test this yourself now - you can
just copy and paste from your script into the command line to make it
easier.</p>
<p>After we have tested the loop to make sure it is working properly,
all we have to do is add the <code>bwa mem</code> command we made
earlier but with our declared variables in place.</p>
<pre class="shell"><code>for IND in ${INDS[@]};
do
    # declare variables
    FORWARD=/home/data/reads/${IND}_R1_001.fastq.gz
    REVERSE=/home/data/reads/${IND}_R2_001.fastq.gz
    OUTPUT=~/align/${INDS%%_*}_sort.bam

    # then align and sort
    echo &quot;Aligning $IND with bwa&quot;
    bwa mem -M -t 4 $REF $FORWARD \
    $REVERSE | samtools view -b | \
    samtools sort -T ${IND} &gt; $OUTPUT

done</code></pre>
<p>With this completed <code>for</code> loop, we have a command that
will create variables, align and sort our reads and also echo to the
screen which individual it is working on, so that we know how well it is
progessing.</p>
<p>Although it takes a bit more work to make a script like this, it is
worth it. This script is very general - you would only need to edit the
variables in order to make it work on almost any dataset on any cluster.
Reusability (and clarity) of scripts is something to strive for.</p>
<p>Now thatwe have built our script, it is time to use it. Save it and
name it <code>align_sort.sh</code> We will move it on to the cluster,
either using <code>scp</code> or <code>filezilla</code>, a process
explained in <a href="">this tutorial</a>. Be sure to move it to your
<code>home</code> folder.</p>
<p>Once the script is on the cluster, open a screen and call it
<code>align</code>; (i.e. <code>screen -S align</code>). Then run the
script like so:</p>
<pre class="shell"><code>bash align_sort.sh</code></pre>
<p>You will now see the script running as the sequences align. Press
<code>Ctrl + A + D</code> in order to leave the screen. Now is a good
time to take a break as you wait for the job to complete.</p>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>





</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
