<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Variant calling</title>

<script src="site_libs/header-attrs-2.25/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">tirupati_genomics</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/markravinet/tirupati_genomics">
    <span class="fab fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Variant calling</h1>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-01-22
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 2
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>tirupati_genomics/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcommarkravinettirupatigenomicstree3f129fa1e4eb0a6e9372205c7b1edd4aa827b8a0targetblank3f129faa">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/markravinet/tirupati_genomics/tree/3f129fa1e4eb0a6e9372205c7b1edd4aa827b8a0" target="_blank">3f129fa</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcommarkravinettirupatigenomicstree3f129fa1e4eb0a6e9372205c7b1edd4aa827b8a0targetblank3f129faa"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/markravinet/tirupati_genomics/tree/3f129fa1e4eb0a6e9372205c7b1edd4aa827b8a0" target="_blank">3f129fa</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/

Untracked files:
    Untracked:  website_setupt.R

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/4_variant_calling.Rmd</code>) and
HTML (<code>docs/4_variant_calling.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/markravinet/tirupati_genomics/blob/3f129fa1e4eb0a6e9372205c7b1edd4aa827b8a0/analysis/4_variant_calling.Rmd" target="_blank">3f129fa</a>
</td>
<td>
Mark Ravinet
</td>
<td>
2024-01-22
</td>
<td>
update files
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>If everything has worked correctly up to this point, we now have a
set of sequence reads that are aligned to our reference genome and
stored as bam files. What we want to do now is to call variants from
these alignments.</p>
<p><strong>Disclaimer: There are many ways to perform variant calling
for short read data - bcftools, GATK, FreeBayes, Stacks
(RAD-seq)</strong> We will be running through a single example today but
it you should evaluate which tool will be best for your purposes with
your own data.</p>
<p>For this tutorial, we will use <code>bcftools</code> which is
designed by the same team behind <code>samtools</code> - <a
href="http://www.htslib.org/">they are part of the same pipeline</a>.
<code>bcftools</code> is itself a comprehensive pipeline and produces a
variant call format (VCF) that is used in many downstream analyses.</p>
<div id="indexing-the-reference-again" class="section level3">
<h3>Indexing the reference… again</h3>
<p>The first thing we need to do is index our reference genome again.
This actually needs to be done with <code>samtools</code>. Return to the
home directory and perform the following actions</p>
<pre class="shell"><code>cd ~/reference/
samtools faidx house_sparrow_genome_assembly-18-11-14_masked.fa</code></pre>
<p>This will create a fasta index, denoted by the <code>.fai</code>
suffix. Let’s take a closer look at that index.</p>
<pre class="shell"><code>head house_sparrow_genome_assembly-18-11-14_masked.fa.fai</code></pre>
<p>You don’t need to worry about this in great deal but for clarity, the
columns are: chromsome name, chromosome length, offset of the first
base, fasta line length, fasta line length + 1.</p>
</div>
<div id="calling-variants" class="section level3">
<h3>Calling variants</h3>
<p>To call variants, we will first use the <code>samtools mpileup</code>
tool to pileup our BAM files. What does this mean exactly? Well we will
take all reads at a given position and call variants from the reads
covering that position. We need to do this for all individuals.</p>
<p>After performing the pileup, we than pass the output to
<code>bcftools call</code> which will actually call variants. To see the
options available to each part of the pipeline, just type their names
into the command line.</p>
<p>Rather than perform each step previously as we did before, here we
will use pipes - <code>|</code> - to pass data from each part of the
pipeline to the next. Rather than explaining every point now, it’s best
we just perform the calling and break it down later.</p>
<p>First of all, use <code>screen</code> to make a screen called variant
and move into it.</p>
<p>Inside the screen, declare a variable for the reference genome.</p>
<pre class="shell"><code>REF=~/reference/house_sparrow_genome_assembly-18-11-14_masked.fa</code></pre>
<p>Next we run the <code>bcftools mpileup</code> and
<code>bcftools call</code> command. We will break it down after.</p>
<pre class="shell"><code>cd ~
bcftools mpileup -a AD,DP,SP -Ou -f $REF \
./align/*_sort.bam | bcftools call -f GQ,GP \
-mO z -o ./sparrows.vcf.gz</code></pre>
<p>While this is running, let’s go through the options and get an idea
of what we did.</p>
<p>For <code>bcftools mpileup</code>:</p>
<ul>
<li><code>-a</code> - Annotate the vcf - here we add allelic depth
(<code>AD</code>), genotype depth (<code>DP</code>) and strand bias
(<code>SP</code>).</li>
<li><code>-O</code> - the output type. Here it is <code>u</code> which
means we do not compress the output.</li>
<li><code>-f</code> - specify the reference genome to call variants
against.</li>
</ul>
<p>For <code>bcftools call</code>:</p>
<ul>
<li><code>-f</code> - format fields for the vcf - here they are genotype
quality (<code>GQ</code>) and genotype probability
(<code>GP</code>).</li>
<li><code>-v</code> - output variant sites only - i.e. ignore
non-variant parts of the reads</li>
<li><code>-m</code>- use bcftools multiallelic caller</li>
<li><code>-O</code>- specify the output type, here it is <code>z</code>
- i.e. gzipped (compressed) vcf</li>
<li><code>-o</code> output path</li>
</ul>
<p>In essence, we have piled up each read against each genome position
and then called variants where there are enough alternative reads to
support the presence of a variant.</p>
<p>If the command is still running (it takes some time), detach your
screen using <code>ctrl + a + d</code>. We will return to our VCF
later.</p>
<div id="exploring-vcf-files" class="section level4">
<h4>Exploring vcf files</h4>
<p>First, a bit of housekeeping. Make a vcf directory and move the vcf
into it.</p>
<pre class="shell"><code>mkdir vcf
mv sparrows.vcf.gz ./vcf
cd vcf</code></pre>
<p>vcf stands for ‘variant call format’ and is a standard format used
for variant calling and in population genomics. Again a detailed
specification can be found <a
href="https://samtools.github.io/hts-specs/VCFv4.2.pdf">online</a>. It
can take a bit of getting used to but it is widely supported and very
useful. Let’s have a look at the vcf.</p>
<pre class="shell"><code>bcftools view -h sparrows.vcf.gz</code></pre>
<p>A lot of information will flash by - this is the vcf header. Like the
SAM file, the header contains information on what has been done to the
vcf. The last line is particularly important as it shows what each field
in the main body of the vcf is and it also gives the individual names.
Try the following:</p>
<pre class="shell"><code>bcftools view -h sparrows.vcf.gz | head
bcftools view -h sparrows.vcf.gz | tail</code></pre>
<p>Here <code>-h</code> means show me only the header. You can also see
<code>-H</code> to see only the raw calls. For some more information on
the vcf format, see here.</p>
<p>A nice feature of vcf files is that you can access almost any part of
the genome you are interested in. To do this though, you need to index
the vcf first. Simply do the following</p>
<pre class="shell"><code>bcftools index sparrows.vcf.gz</code></pre>
<p>Let’s see what variants are present at the start of chr2:</p>
<pre class="shell"><code>bcftools view -H -r chr2 sparrows.vcf.gz | head</code></pre>
<p>Here the <code>-r</code> flag specfies the region of the genome to
examine. We can achieve the same effect by specifying the base pair
coordinates</p>
<pre class="shell"><code>bcftools view -H -r chr2:100000-200000 sparrows.vcf.gz</code></pre>
<p>Now you’re probably wondering what exactly all this data actually
means. Let’s explore in a bit more detail.</p>
<pre class="shell"><code>bcftools view -h sparrows.vcf.gz | tail -1 | cut -f 1-9
bcftools view -H sparrows.vcf.gz | head -1 | cut -f 1-9</code></pre>
<p>These are the first 11 fields of the vcf and they are always present.
What do they mean?</p>
<ul>
<li><code>CHROM</code> - chromosome or scaffold id from the reference
genome</li>
<li><code>POS</code> - base pair reference position</li>
<li><code>ID</code> - SNP id - blank in this case</li>
<li><code>REF</code>- Reference base - A,C,G,T or N. More than one
indicates an indel</li>
<li><code>ALT</code> - Alternaate base - the alternate base called as a
variant</li>
<li><code>QUAL</code> - Phred quality score for alternate base call
(site not individual)</li>
<li><code>FILTER</code> - filter status - if PASS then all filters
passed</li>
<li><code>INFO</code> - additional info on each site - explanation
stored in header</li>
<li><code>FORMAT</code>- the format for the genotype fields</li>
</ul>
<p>There is a lot of information here! Don’t worry too much if it
doesn’t make perfect sense immediately - familiarity with this format
will come with time and experience.</p>
<p>Let’s look at the the first site again in detail:</p>
<pre class="shell"><code>bcftools view sparrows.vcf.gz | grep -m 1 -A 1 &quot;#CHROM&quot; | cut -f 1-7</code></pre>
<p>You should see this:</p>
<pre class="shell"><code>#CHROM  POS     ID      REF     ALT     QUAL    FILTER
chr1    1046    .       C       T       19.9754 .</code></pre>
<p>Which means we have a variant at 1046 on chr1. The reference base is
C, alternate base is T. There are no filters.</p>
<p>Have a look at the info field with the following code</p>
<pre class="shell"><code>bcftools view -H sparrows.vcf.gz | head -1 | cut -f 8</code></pre>
<p>You can find out what these mean by grepping the header.</p>
<pre class="shell"><code>bcftools view -h sparrows.vcf.gz | grep &quot;INFO&quot;</code></pre>
<p>So for example, DP here means the raw read depth for the entire
site.</p>
<p>What about the actual genotype information? Firstly it is wise to
look at the format here.</p>
<pre class="shell"><code>bcftools view -H sparrows.vcf.gz | head -1 | cut -f 9</code></pre>
<p>To investigate what these mean, grep the header again.</p>
<pre class="shell"><code>bcftools view -h sparrows.vcf.gz | grep &quot;##FORMAT&quot;</code></pre>
<p>Now let’s take a look at the call for a single individual.</p>
<pre class="shell"><code>bcftools view -H sparrows.vcf.gz | head | cut -f 20</code></pre>
<p>This will return:</p>
<pre class="shell"><code>1/1:41,3,0:1:0,1,0:65,12,0:12</code></pre>
<p>First we have the genotype. Here <code>0</code> always denotes the
reference, <code>1</code> the alternate base so we can see this
individual is homozygous for the <code>ALT</code> base. We will ignore
the genotype likelhood for now. You can see there is 1 read in total
covering this site - which we are seeing because we used a reduced
dataset.</p>
<p>Is there an easier way to view genotypes? Yes there is - using the
<code>bcftools query</code> utility.</p>
<p>This allows you to look at the genotypes like so:</p>
<pre><code>bcftools query -f &#39;%CHROM\t%POS\t%REF\t%ALT[\t%GT]\n&#39; sparrows.vcf.gz | head</code></pre>
<p>You can also translate them into actual basecalls. Try this:</p>
<pre><code>bcftools query -f &#39;%CHROM\t%POS\t%REF\t%ALT[\t%TGT]\n&#39; sparrows.vcf.gz | head</code></pre>
<p>The <code>bcftools query</code> utility is very powerful and a useful
tool to know about for file conversion in your own work.</p>
<p>Looking at the genotypes, you will see they are mostly missing. One
of the reasons for this is the fact we used such a tiny dataset. So in
the next section we will look at a proper set of variant calls from
multiple individuals.</p>
<hr />
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>





</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
